segmentation_models_pytorch/encoders/sam.py

```diff
@@ -0,0 +1,106 @@
+import math
+import warnings
+from typing import Mapping, Any
+
+import torch
+from segment_anything.modeling import ImageEncoderViT
+
+from segmentation_models_pytorch.encoders._base import EncoderMixin
+
+
+class SamVitEncoder(EncoderMixin, ImageEncoderViT):
+    def __init__(self, **kwargs):
+        self._vit_depth = kwargs.pop("vit_depth")
+        self._encoder_depth = kwargs.get("depth", 5)
+        kwargs.update({"depth": self._vit_depth})
+        super().__init__(**kwargs)
+        self._out_chans = kwargs.get("out_chans", 256)
+        self._patch_size = kwargs.get("patch_size", 16)
+        self._validate()
+
+    @property
+    def output_stride(self):
+        return 32
+
+    def _get_scale_factor(self) -> float:
+        """Input image will be downscale by this factor"""
+        return int(math.log(self._patch_size, 2))
+
+    def _validate(self):
+        # check vit depth
+        if self._vit_depth not in [12, 24, 32]:
+            raise ValueError(f"vit_depth must be one of [12, 24, 32], got {self._vit_depth}")
+        # check output
+        scale_factor = self._get_scale_factor()
+        if scale_factor != self._encoder_depth:
+            raise ValueError(
+                f"With patch_size={self._patch_size} and depth={self._encoder_depth}, "
+                "spatial dimensions of model output will not match input spatial dimensions. "
+                "It is recommended to set encoder depth=4 with default vit patch_size=16."
+            )
+
+    @property
+    def out_channels(self):
+        # Fill up with leading zeros to be used in Unet
+        scale_factor = self._get_scale_factor()
+        return [0] * scale_factor + [self._out_chans]
+
+    def forward(self, x: torch.Tensor) -> list[torch.Tensor]:
+        # Return a list of tensors to match other encoders
+        return [x, super().forward(x)]
```

Comment:
I still didnt get how Unet works with just two feature tensors, probably we need some intermediate layers + upsampling/downsampling for proper architecture, see this PR for ViT adaptation

https://github.com/qubvel/segmentation_models.pytorch/pull/776

Reply:
I'm keeping the first one just because Unet decoder needs at least two tensors from the encoder.
The decoder uses skip connections to concat them to each upsample block. Just in this case it does not concatenate anything as there's no corresponding encoder output. I'm not sure we can have skip connections here because of multiple reasons:
1. Vit depth is pretty large: 12, 24 or 32. 
2. Each Transformer block has the same output of `[batch_size, out_chans, H//16, W//16]`, whereas for the Unet decoder we need encoder outputs to halve H/W at each stage.

I could look however how Timm does it, Maybe I could figure out something.

Reply:
We can add upsampling/downsampling steps
e.g. we got intermediate outputs from ViT

```python
features = [
  x
  [batch_size, out_chans, H//16, W//16]  # vit_layer 5, bilinear upsample x8
  [batch_size, out_chans, H//16, W//16]  # vit_layer 10, bilinear upsample x4
  [batch_size, out_chans, H//16, W//16]  # vit_layer 15, bilinear upsample x2
  [batch_size, out_chans, H//16, W//16]  # vit_layer 20, as is
  [batch_size, out_chans, H//16, W//16]  # vit_layer last, bilinear downsample x2
]
```

This is a dummy example, I guess this should work better than just the last feature output